### PR TITLE
valeStyles: init

### PR DESCRIPTION
**Pull request by marsam**
## Description of changes

Add valeStyles packages

cc @katexochen 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/tools/text/vale/default.nix [hunk 1]
```diff
@@ -26,6 +34,16 @@ buildGoModule rec {
...
+    paths = [ vale ] ++ selector valeStyles;
+    nativeBuildInputs = [ makeBinaryWrapper ];
+    postBuild = ''
+      wrapProgram "$out/bin/vale" \
+        --set VALE_STYLES_PATH "$out/share/vale/styles/"
+    '';
+  };
+
   meta = with lib; {
     description = "A syntax-aware linter for prose built with speed and extensibility in mind";
```
Comment (katexochen) [inline pkgs/tools/text/vale/default.nix:48]: Maybe we could add a `longDescription` to vale hinting how to use vale with styles?

File: pkgs/tools/text/vale/default.nix [hunk 2]
```diff
@@ -26,6 +34,16 @@ buildGoModule rec {
   # Tests require network access
   doCheck = false;
 
+  passthru.withStyles = selector: symlinkJoin {
```
Comment (katexochen) [inline pkgs/tools/text/vale/default.nix:37]: In the `postInstall` of vale, we are currently copying the styles from `/testdata` into a separate `data` output. The [README of that dir](https://github.com/errata-ai/vale/blob/v3/testdata/styles/README.md) says the styles are for testing only and should not be used. Now that we have the real styles packaged, should we just remove the `data` output?

**Review by katexochen — CHANGES_REQUESTED** [review]
Thanks a lot, works perfectly fine!

```
❯ result/bin/vale ls-vars
Variable         | Description                                                         | Value
VALE_CONFIG_PATH | Override the default search process by specifying a .vale.ini file. | ✗
VALE_STYLES_PATH | Specify the location of the default StylesPath.                     | /nix/store/a27m06an8ydx9mzjsgxw7qhn1kwsxb06-vale-with-styles-3.3.1/share/vale/styles/
```

Just some nits.

File: pkgs/tools/text/vale/styles.nix [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+      owner = "errata-ai";
+      repo = "proselint";
+      rev = "v${version}";
+      hash = "sha256-faeWr1bRhnKsycJY89WqnRv8qIowUmz3EQvDyjtl63w=";
+    };
+    meta.license = lib.licenses.bsd3;
+  };
+
+  readability = buildStyle rec {
+    name = "readability";
```
Comment (katexochen) [inline pkgs/tools/text/vale/styles.nix:83]: Result of `nixpkgs-review pr 301204` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>valeStyles.readability</li>
  </ul>
</details>
<details>
  <summary>7 packages built:</summary>
  <ul>
    <li>valeStyles.alex</li>
    <li>valeStyles.google</li>
    <li>valeStyles.joblint</li>
    <li>valeStyles.microsoft</li>
    <li>valeStyles.proselint</li>
    <li>valeStyles.redhat</li>
    <li>valeStyles.write-good</li>
  </ul>
</details>

```suggestion
    name = "Readability";
```

**Comment by katexochen** [comment]
@marsam any chance we get this in before the release?

File: pkgs/tools/text/vale/default.nix (hunk 1 above)
Comment (marsam) [reply pkgs/tools/text/vale/default.nix:48]: Good idea, I added a small documentation

File: pkgs/tools/text/vale/styles.nix [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    }@args:
+    stdenvNoCC.mkDerivation ({
+      pname = "vale-style-${name}";
+      installPhase = ''
+        runHook preInstall
+        install -Dm644 ${stylePath}/*.yml -t "$out/share/vale/styles/${name}"
+        runHook postInstall
+      '';
+      meta = {
+        platforms = lib.platforms.all;
```
[the remark this answers is not in the document]
Comment (marsam) [reply pkgs/tools/text/vale/styles.nix:27]: unfortunately, I may not be able to maintain these styles.

File: pkgs/tools/text/vale/default.nix (hunk 2 above)
Comment (marsam) [reply pkgs/tools/text/vale/default.nix:31]: Good catch! Yes, it makes sense, I've removed the data output.

**Comment by marsam** [comment]
Hi, sorry for the delay. I had my changes locally, but somehow I forgot to push them :( 

Thanks for reviewing! I addressed you comments, please take a look.

File: pkgs/tools/text/vale/styles.nix (hunk 4 above)
[the remark this answers is not in the document]
Comment (katexochen) [reply pkgs/tools/text/vale/styles.nix:27]: In that case feel free to add me as maintainer for the styles. :)

**Review by katexochen — APPROVED** [review]
Result of `nixpkgs-review pr 301204` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>8 packages built:</summary>
  <ul>
    <li>vale</li>
    <li>valeStyles.alex</li>
    <li>valeStyles.google</li>
    <li>valeStyles.joblint</li>
    <li>valeStyles.microsoft</li>
    <li>valeStyles.proselint</li>
    <li>valeStyles.readability</li>
    <li>valeStyles.write-good</li>
  </ul>
</details>

File: pkgs/tools/text/vale/styles.nix (hunk 4 above)
[the remark this answers is not in the document]
Comment (marsam) [reply pkgs/tools/text/vale/styles.nix:27]: Cool! I added you as a maintainer

**Pull request merged**